source/Octopus.Tentacle/Scripts/RunningScript.cs

```diff
@@ -53,37 +67,51 @@ public void Execute()
                             token,
                             log))
                         {
-                            RunScript(shellPath, writer);
+                            State = ProcessState.Running;
+
+                            if (stateRepository != null)
+                            {
+                                var scriptState = stateRepository.Load();
+                                scriptState.Start();
+                                stateRepository.Save(scriptState);
+                            }
+
+                            exitCode = RunScript(shellPath, writer);
                         }
                     }
                     catch (OperationCanceledException)
                     {
                         writer.WriteOutput(ProcessOutputSource.StdOut, "Script execution canceled.");
-                        ExitCode = ScriptExitCodes.CanceledExitCode;
-                        State = ProcessState.Complete;
+                        exitCode = ScriptExitCodes.CanceledExitCode;
                     }
                     catch (TimeoutException)
                     {
                         writer.WriteOutput(ProcessOutputSource.StdOut, "Script execution timed out.");
-                        ExitCode = ScriptExitCodes.TimeoutExitCode;
-                        State = ProcessState.Complete;
+                        exitCode = ScriptExitCodes.TimeoutExitCode;
                     }
                 }
             }
             catch (Exception)
             {
                 // Something went really really wrong, probably creating or writing to the log file (Disk space)
-                ExitCode = ScriptExitCodes.FatalExitCode;
-                State = ProcessState.Complete;
+                exitCode = ScriptExitCodes.FatalExitCode;
+            }
+
+            if (stateRepository != null)
+            {
+                var scriptState = stateRepository.Load();
+                scriptState.Complete(exitCode);
+                stateRepository.Save(scriptState);
             }
+
+            ExitCode = exitCode;
```

Comment:
Refactor to ensure we dispose of the script writer before we set the exit code and state

Reply:
To be super duper safe we mark the script as complete the ExitCode and state should be put into a try finally